jetty-websocket/websocket-core/src/main/java/org/eclipse/jetty/websocket/core/internal/FrameFlusher.java

```diff
@@ -259,26 +276,47 @@ private int getQueueSize()
         }
     }
 
-    @Override
-    public void succeeded()
-    {
-        succeedEntries();
-        super.succeeded();
-    }
-
-    private boolean succeedEntries()
+    public void timeoutExpired()
     {
-        boolean hadEntries = false;
-        for (Entry entry : entries)
+        boolean failed = false;
+        synchronized (FrameFlusher.this)
         {
-            hadEntries = true;
-            if (entry.frame.getOpCode() == OpCode.CLOSE)
-                endPoint.shutdownOutput();
-            notifyCallbackSuccess(entry.callback);
-            entry.release();
+            if (closedCause != null)
+                return;
+
+            long time = System.currentTimeMillis();
+            long nextTimeout = -1;
+
+            Iterator<Entry> iterator = TypeUtil.concat(entries.iterator(), queue.iterator());
+
+            while (iterator.hasNext())
+            {
+                Entry entry = iterator.next();
+                long timeSinceCreation = time - entry.getTimeOfCreation();
```

Comment:
rather than doing this calculation for every iteration, how about just working out `long expiredIfCreatedBefore = time-idleTimeout;` then do a simple comparison every loop.